### DISCUSSION TITLE
Tide Clock - perpetual tide times for the UK and worldwide

### DISCUSSION THREAD

**Discussion by drtonyr** [wasp-os/wasp-os#483] [Show and tell]
I've pretty much finished my Tide Clock face for wasp-os, you can get it from https://github.com/drtonyr/wasp-os/tree/TideFace.   This is what it looks like:

![TideClockApp](https://github.com/wasp-os/wasp-os/assets/30016701/394587c6-7264-49bc-a23a-abc0a9f1c1f9)

I'm not going to submit a PR because (a) it won't merge to a single commit right now and (b) there's too much to talk about first.

Let's start with what went well.  I'm pretty pleased with the tide engine that sits underneath (and can be found at https://github.com/drtonyr/tide).  This took a lot of work, it's pretty easy to get something that is right to +-40 minutes, but that means being more than 80 minutes out 5% of the time and that wasn't good enough for me.   After a lot of modelling I got it down to +-9 minutes (on average), which I'm happy about as it's a tiny fraction of the tidal range.  However, this needs a database of locations, more on that later.  The engine also needs to work with Python and Micropython, and in micropython we don't get 64bit floats, so there is work to keep things in ints and reasonable rounding errors.  I'm also very happy with the layout, in the space of a normal clock face you get the time and date in the normal fonts plus the location, fraction of tidal range and the times and hours to the next four tides.  That's a lot of information, more than I thought I could fit in.  I'd like some of the text to be left justified, there is #469 for that.

Tides don't know about daylight savings, and neither does micropython.  I wrote some code that will do most of the conversions needed, that is the European and US/Canada regions in https://en.wikipedia.org/wiki/Daylight_saving_time_by_country.

Now for pushing the boundaries a bit.  I needed a way to set location, I decided that going through a list was acceptable and noticed that the down swipe was available from clock faces.   So I commandeered that which works well when there are only a few to choose from.   However, I have lots of locations in the UK and another load to cover worldwide, so you've got to do some hand editing of the config file before pushing it to the watch.  Which brings me to the use of the /flash/data directory, I think we've got a long way in #458 but it's still not finished, in particular files under /flash/data need to be distributed with git but if the user then needs to edit what happens with updates?  (this may well have a git solution like gitignore, but git and I fight too much).

Finally, the supplied config file is in very simple csv format which is read line by line.  We have ujson available, but there just isn't the RAM to use it.  We seem chronically short on RAM in the standard  distribution, I'd love to have time to find out where it's all going, safe mode has lots of RAM free.

One last point.   This watch face, like many, uses a self._min variable so that nothing changes until the minute changes.  This isn't set in __init__ and I've seen exceptions thrown when this has been read before set.  It could be bugs in my code only, it could be a race condition in all watch faces that use _min, I don't know.   Maybe we should reorganise so that there's a general clock face that all specific clock faces are a subclass of, I'm really not enough of a python expert to know the work here.

In summary, I started by writing one thing and ended up writing three which I've tried to keep separate.   There's a tide engine, there's a clock face and there is compensation for daylight savings.  It does what I want it to do, but it raises all sorts of issues so I'm not pushing for inclusion in wasp-os right now.
